### ISSUE TITLE
Got bad result when implementing Transfer learning on CACD pre-trained model 

### ISSUE THREAD

**Issue by fefei69** [Raschka-research-group/coral-cnn#39]
I tried to fine tune the CACD pre-trained model (e.g : cacd-coral_seed1) on my small dataset (training:121 validation : 15 testing :15) age range: 52~93 , 42 classes. The training result is like this: MAE/RMSE: | Best Train: 1.44/2.77 | Best Valid: 4.00/6.43 | Best Test: 5.07/6.68. The MAE is not bad but when i plot the error(estimated age - real age) v.s real age , i found that the error decreased when the real age increased, meaning that the error and the real age are correlated. I don't think that is normal. I think that indicates the model can't inference properly and just output the same age so the error (estimated age - real age)  is positive when the real age is small and negative when the real age is large. I have no idea what kind of problems might cause this. Maybe something went wrong with my code ? or my dataset is just too small to get a good model?
### Adjustment in ResNet (freeze all the layers before layer4):

```
class ResNet(nn.Module):
    def __init__(self, block, layers, num_classes, grayscale):
        self.num_classes = num_classes
        self.inplanes = 64
        if grayscale:
            in_dim = 1
        else:
            in_dim = 3
        super(ResNet, self).__init__()
        self.conv1 = nn.Conv2d(in_dim, 64, kernel_size=7, stride=2, padding=3,
                            bias=False)
        self.bn1 = nn.BatchNorm2d(64)
        self.relu = nn.ReLU(inplace=True)
        self.maxpool = nn.MaxPool2d(kernel_size=3, stride=2, padding=1)
        self.layer1 = self._make_layer(block, 64, layers[0]) # stride預設1 不會改變圖片大小
        self.layer2 = self._make_layer(block, 128, layers[1], stride=2)
        self.layer3 = self._make_layer(block, 256, layers[2], stride=2)
        #self.layer4 = self._make_layer(block, 512, layers[3], stride=2)
        #self.avgpool = nn.AvgPool2d(4)
        #self.fc = nn.Linear(512, 1, bias=False)
        # self.linear_1_bias = nn.Parameter(torch.zeros(self.num_classes-1).float())

        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                n = m.kernel_size[0] * m.kernel_size[1] * m.out_channels
                m.weight.data.normal_(0, (2. / n)**.5)
            elif isinstance(m, nn.BatchNorm2d):
                m.weight.data.fill_(1)
                m.bias.data.zero_()
```

### Replace the new layers that i would like to train :

```
#resnet34 setting
layers=[3, 4, 6, 3]
block=BasicBlock
#model.layer3 = model._make_layer(block, 256, layers[2], stride=2)
model.layer4 = model._make_layer(block, 512, layers[3], stride=2)
model.avgpool = nn.AvgPool2d(4)
model.fc = nn.Linear(512, 1, bias=False)
model.linear_1_bias = nn.Parameter(torch.zeros(NUM_CLASSES-1).float())
for m in model.modules():
    if isinstance(m, nn.Conv2d):
        n = m.kernel_size[0] * m.kernel_size[1] * m.out_channels
        m.weight.data.normal_(0, (2. / n)**.5)
    elif isinstance(m, nn.BatchNorm2d):
        m.weight.data.fill_(1)
        m.bias.data.zero_()

model.to(DEVICE)
#print trainable parameters
print("Params to learn:")
feature_extract = True
if feature_extract:
    params_to_update = []
    for name,param in model.named_parameters():
        if param.requires_grad == True:
            params_to_update.append(param)
            print("\t",name)
else:
    for name,param in model.named_parameters():
        if param.requires_grad == True:
            print("\t",name)
optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
```

**Comment by rasbt** [comment]
Interesting! I think there is a massive amount of overfitting due to the small dataset size: 

- Train: 1.44/2.77 | 
- Best Valid: 4.00/6.43 | 
- Best Test: 5.07/6.68.

I think the model just memorizes the training examples. If you have a large model like ResNet, it's very easy to overfit on small datasets. 
Have you considered just using a pre-trained model and just finetuning the last layer on your dataset?

**Comment by fefei69** [comment]
Thank you so much for your reply! I had tried to fine-tune the last layer before (train only `model.fc = nn.Linear(512, 1, bias=False)` and `model.linear_1_bias = nn.Parameter(torch.zeros(NUM_CLASSES-1).float())` ) but the result was worse than before, something like 
MAE/RMSE: | Best Train: 6.35/8.14 | Best Valid: 4.60/6.30 | Best Test: 6.13/7.96. 
I'm guessing that maybe something went wrong with my labeling. Since the labels of pre-trained model say morph2 model is 0-54 correspond to ages 16-70 but my dataset's labels is 0-41 correspond to ages 52-93. Therefore, the model might confuse because the label 0 corresponded to age 16 but now corresponds to age 52. That may explain why i finetuning the last layer only gets worse result than i trained more layers(last few layers). Do you have any idea on that or any solutions? Thanks in advance for your reply!

**Comment by rasbt** [comment]
I see. Yeah, the labeling could be an issue here. But honestly, the Morph-2 dataset is also taken under very specific lighting conditions and so forth, so it's probably that the model is really hard to generalize to other face image datasets, even with fine-tuning. It'd probably be better to pre-train the model on a dataset that is maybe more similar to your dataset. Perhaps something like CACD.

**Comment by fefei69** [comment]
Sorry for the late reply. I had tried to pre-train AFAD, CACD and MORPH-2 but still getting bad results. I think the problem is that my dataset is just too small( 151 images) so I won't get a proper model no matter how I adjust the parameters or the pre-trained model. I might - 1.try to augment my data through adjusting the pixel intensity of the images to generate more data. 
2. Since my goal is to estimate the people whose age is between around 50-90 years old, so i might use CACD as pre-trained model and fine-tune it on UTKFace dataset (only the people aged between 50-90).
3. train from scratch on old people dataset(haven't found a proper one).
I think 2. is most likely to get a proper model. Do you have any thoughts about my methods or any suggestions for the specific goal (Age estimation for elderly people)? Thank you so much for your patience!

**Comment by rasbt** [comment]
Arg, sorry to hear. But yeah, like you suggest, the dataset is likely too small. Image augmentation might help a bit, but I wouldn't expect miracles. Self-supervised learning -- if you have lots of unlabeled images -- can also help somewhat. The probably best bang for the buck is consistent preprocessing though, making sure that the faces are all centered and the same size etc.
